### ISSUE TITLE
Recursive JSON object type hint not working

### ISSUE THREAD

**Issue by jnooree** [microsoft/pyright#3321]
**Describe the bug**
Recursive JSON object type hinting does not work. Pyright instead reports `_T@list is invariant` diagnostic.

**To Reproduce**
```python
from typing import List, Dict, Tuple, Union

SerializablePri = Union[str, int, float, bool, None]
SerializableVal = Union[SerializablePri, "SerializableObj"]
SerializableDic = Dict[str, SerializableVal]
SerializableSeq = Union[List[SerializableVal], Tuple[SerializableVal, ...]]
SerializableObj = Union[SerializableDic, SerializableSeq]


def test(obj: SerializableObj):
    pass


obj = [{"a": 1, "b": 2}, {"c": 3, "d": 4}]  # list[dict[str, int]]
test(obj)  # Diagnosed as:
# Argument of type "list[dict[str, int]]" cannot be assigned to parameter "obj" of type "SerializableObj" in function "test"
#   Type "list[dict[str, int]]" cannot be assigned to type "SerializableObj"
#     "list[dict[str, int]]" is incompatible with "SerializableDic"
#     TypeVar "_T@list" is invariant
#       Type "dict[str, int]" cannot be assigned to type "SerializableVal"
#         "dict[str, int]" is incompatible with "str"
#     "list[dict[str, int]]" is incompatible with "Tuple[SerializableVal, ...]"
```

**Expected behavior**
Must accept such argument without diagnostics.

**Screenshots or Code**
Provided.

**VS Code extension or command-line**
On python `v3.9.12`, pylance `v2022.4.0`, vscode `v1.66.1`, macOS `v12.3.1`.

**Additional context**
N/A.

**Comment by erictraut** [comment]
The behavior is correct. The type parameter for `list` is invariant, so you can't assign a `list[dict[str, int]]` to a `List[SerializableVal]`.

You have a few options:
1. Switch from `list` to `Sequence` (and `dict` to `Mapping`) because the type parameters for `Sequence` and `Mapping` are covariant instead of invariant.
2. Use an explicit type annotation for `obj` so [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types) can infer the correct type for your literal list expression.

```python
obj: SerializableObj = [{"a": 1, "b": 2}, {"c": 3, "d": 4}]
```
